### PR TITLE
win: more robust parsing of SDK version

### PR DESCRIPTION
**Pull request by refack**
Fixes: https://github.com/nodejs/node-gyp/issues/1179

Personally I prefer #1178, but this is smaller.

### CODE REVIEW

File: lib/Find-VS2017.cs [hunk 1]
```diff
@@ -228,9 +228,10 @@ private static bool CheckInstance(ISetupInstance2 setupInstance2, ref StringBuil
                     hasMSBuild = true;
                 else if (id == "Microsoft.VisualStudio.Component.VC.Tools.x86.x64")
                     hasVCTools = true;
-                else if (id.StartsWith(Win10SDKPrefix))
-                    Win10SDKVer = Math.Max(Win10SDKVer, UInt32.Parse(id.Substring(Win10SDKPrefix.Length)));
-                else if (id == "Microsoft.VisualStudio.Component.Windows81SDK")
+                else if (id.StartsWith(Win10SDKPrefix)) {
+                    string[] parts = id.ToString().Substring(Win10SDKPrefix.Length).Split('.');
```
Comment (bnoordhuis) [inline lib/Find-VS2017.cs]: Superfluous ToString() call?

Comment (refack) [inline lib/Find-VS2017.cs]: Yep.
Gone.

File: lib/Find-VS2017.cs [hunk 2]
```diff
@@ -228,9 +228,10 @@ private static bool CheckInstance(ISetupInstance2 setupInstance2, ref StringBuil
                     hasMSBuild = true;
                 else if (id == "Microsoft.VisualStudio.Component.VC.Tools.x86.x64")
                     hasVCTools = true;
-                else if (id.StartsWith(Win10SDKPrefix))
-                    Win10SDKVer = Math.Max(Win10SDKVer, UInt32.Parse(id.Substring(Win10SDKPrefix.Length)));
-                else if (id == "Microsoft.VisualStudio.Component.Windows81SDK")
+                else if (id.StartsWith(Win10SDKPrefix)) {
+                    string[] parts = id.Substring(Win10SDKPrefix.Length).Split('.');
+                    Win10SDKVer = Math.Max(Win10SDKVer, UInt32.Parse(parts[0]));
```
Comment (joaocgreis) [inline lib/Find-VS2017.cs]: Only the Desktop C++ SDK is good, according to https://docs.microsoft.com/en-us/visualstudio/install/workload-component-id-vs-community it has `.Desktop` appended. What about:

```patch
-                    Win10SDKVer = Math.Max(Win10SDKVer, UInt32.Parse(parts[0]));
+                    if (parts.Length == 1 || (parts.Length == 2 && parts[1] == "Desktop"))
+                        Win10SDKVer = Math.Max(Win10SDKVer, UInt32.Parse(parts[0]));
+                    else
+                        continue;
```

Comment (refack) [inline lib/Find-VS2017.cs]: Ack, with variation

File: lib/Find-VS2017.cs [hunk 3]
```diff
@@ -228,9 +228,12 @@ private static bool CheckInstance(ISetupInstance2 setupInstance2, ref StringBuil
                     hasMSBuild = true;
                 else if (id == "Microsoft.VisualStudio.Component.VC.Tools.x86.x64")
                     hasVCTools = true;
-                else if (id.StartsWith(Win10SDKPrefix))
-                    Win10SDKVer = Math.Max(Win10SDKVer, UInt32.Parse(id.Substring(Win10SDKPrefix.Length)));
-                else if (id == "Microsoft.VisualStudio.Component.Windows81SDK")
+                else if (id.StartsWith(Win10SDKPrefix)) {
+                    string[] parts = id.Substring(Win10SDKPrefix.Length).Split('.');
+                    if (parts.Length == 2 && parts[1] != "Desktop")
```
Comment (joaocgreis) [inline lib/Find-VS2017.cs]: There is already a `Microsoft.VisualStudio.Component.Windows10SDK.15063.UWP.Native` that this does not catch

Comment (refack) [inline lib/Find-VS2017.cs]: 👎 for Microsoft.
Enhanced the guard.

**Comment by joaocgreis** [comment]
CI: https://ci.nodejs.org/view/All/job/nodegyp-test-pull-request/13/

@refack can you land here or should I? (we should still wait for the 72h)

**Comment by refack** [comment]
@joaocgreis I don't have write here (yet?), after we land we need to ping one of the package owners to package a patch version.

**Comment by joaocgreis** [comment]
v0.12 failed, here is another run: https://ci.nodejs.org/job/nodegyp-test-commit/113/ . This one passed.

**Comment by refack** [comment]
Landed in bad903a

**Comment by refack** [comment]
Post land CI: https://ci.nodejs.org/job/nodegyp-test-pull-request/14/

**Comment by refack** [comment]
This should go into `v3.6.2` fixes #1179 and https://github.com/nodejs/node-gyp/issues/1144#issuecomment-301213979
/cc @bnoordhuis @rvagg

**Comment by refack** [comment]
/cc @rvagg @bnoordhuis @joaocgreis @nodejs/node-gyp 
IMHO we need to release another patch version for this (and try to push it into `npm@5`)
Fixes: https://github.com/nodejs/node-gyp/issues/1179
Ref: https://github.com/nodejs/node-gyp/issues/1056#issuecomment-304496936
Ref: https://github.com/nodejs/node-gyp/issues/1144#issuecomment-294879198
Ref: https://github.com/nodejs/node-gyp/pull/1205

**Comment by rvagg** [comment]
:+1: done, v3.6.2, thanks @refack

**Comment by refack** [comment]
> 👍 done, v3.6.2, thanks @refack

No, thank you!

**Pull request merged**